compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_combine_value_barriers.mlir

```diff
@@ -0,0 +1,129 @@
+// RUN: iree-opt --iree-codegen-gpu-combine-value-barriers %s --split-input-file | FileCheck %s
+
+// Since the pass only rearanges the order of instructions, we only check the
+// number of value_barriers.
+
+func.func @tensor_barrier(%write: vector<8xf16>, %input: tensor<8xf16>, %input2 : tensor<16xf16>) -> (vector<8xf16>, vector<8xf16>) {
+  %c0 = arith.constant 0 : index
+  %cv0 = arith.constant 0.0 : f16
+
+  %wait1 = vector.transfer_write %write, %input[%c0] : vector<8xf16>, tensor<8xf16>
+  %synced1 = iree_gpu.value_barrier %wait1 : tensor<8xf16>
+  %out = vector.transfer_read %synced1[%c0], %cv0 : tensor<8xf16>, vector<8xf16>
+
+  %wait2 = vector.transfer_write %write, %input2[%c0] : vector<8xf16>, tensor<16xf16>
+  %synced2 = iree_gpu.value_barrier %wait2 : tensor<16xf16>
+  %out2 = vector.transfer_read %synced2[%c0], %cv0 : tensor<16xf16>, vector<8xf16>
+
+  return %out, %out2 : vector<8xf16>, vector<8xf16>
+}
+
+// There should be only 1 value_barrier left
+
+// CHECK-LABEL: func.func @tensor_barrier
+// CHECK: value_barrier
+// CHECK-NOT: value_barrier
+
+// -----
+
+func.func @vector_barrier(%write: vector<8xf16>, %write2: vector<8xf16>) -> vector<8xf16> {
+  %synced  =  iree_gpu.value_barrier %write  : vector<8xf16>
+  %synced2 =  iree_gpu.value_barrier %write2 : vector<8xf16>
+  %add     =  arith.addf %synced, %synced2   : vector<8xf16>
+  return %add : vector<8xf16>
+}
+
+// There should be only 1 value_barrier left
+
+// CHECK-LABEL: func.func @vector_barrier
+// CHECK: value_barrier
+// CHECK-NOT: value_barrier
```

Comment:
shouldn't this fail because you are (currently) returning failure if both ops don't have tensor semantics?